### PR TITLE
Remove stestr's grouping configuration

### LINKED ISSUE #6105: Use unique filenames to allow test parallelization
Our test suite mostly passes when run with pytest-xdist parallelization, giving a pretty nice speed boost.
However there are just a couple places where tests try to reuse the same temp file and step on each others' toes. This can be avoided by `--dist=loadscope` but that seems a bit fragile and is one more thing to remember.

I know that we can also parallelize using stestr but that is probably unfamiliar to most contributors, and doesn't integrate with tools such as vs code's testing tab (which does know about pytest-xdist). I assume stestr does something similar to `dist=loadscope` by default which is why this doesn't come up for stestr? Or is it just luck about the way stestr does the distribution to workers?

The offending spots that I noticed are in `TestGateMap` and `TestUserConfig` and it seems like replacing the hardcoded filenames with appropriate calls to `tempfile.TemporaryFile` etc would be a pretty straightforward fix.

### PR DESCRIPTION
**Pull request by jorconnor**
Removing  the grouping regex in .stestr.conf should allow stestr to pack its test workers better, and therefore improve the speed of test execution. However, pre-existing test cases utilizing non-unique tempfiles began to fail after the configuration change. This commit modifies such tests to pass under these new conditions. This also allows the test suite to properly execute using pytest-xdist.

Fixed #6105

### Summary

Per #6105 a few tests (TestGateMap and TestUserConfig) were failing when using pytest-xdist.  Both tests were reliant on a hardcoded file name "temp.txt", which is the root source of the failures.  This pull modifies both to use less brittle file buffers. 

I also went ahead and removed the regex found in https://github.com/Qiskit/qiskit-terra/blob/master/.stestr.conf#L3 as discussed in the issue. After correcting the tests I have not seen any issues with the configuration change. I will adjust the merge request if any reviewers are uncomfortable removing it or if it fails in ci checks.

### Details and comments

For TestGateMap:
I modified the test so that it utilizes ```io.BytesIO``` instead of needlessly writing to the problamatic temp file.

For TestUserConfig:
I moved the "string" configurations to their own personal files found in test/python/configurations. This might make it a bit harder to tell what the test is testing i.e. in the future devs will have to open the configuration file. However, this avoids the write to the problematic temp file and should create a less brittle test.

### CODE REVIEW

File: test/python/test_user_config.py [hunk 1]
```diff
@@ -20,128 +20,52 @@
...
 class TestUserConfig(QiskitTestCase):
-    @classmethod
-    def setUpClass(cls):
-        cls.file_path = "temp.txt"
+    def setUp(self):
+        super().setUp()
+        current_dir = os.path.dirname(os.path.abspath(__file__))
+        test_name = str(self.id()).split(".")[-1]
+        file_path = os.path.join(current_dir, "configurations", test_name + ".conf")
+        self.config = user_config.UserConfig(file_path)
```
Comment (mtreinish) [inline test/python/test_user_config.py:28]: I'm not super thrilled about this approach. Personally I find this kind of fragile to have to match the test name with a filename in a relative directory. Also as you mentioned it moves the configuration away from the test code which makes it harder to develop and debug tests here. I do wonder if there is an approach we can use that avoids shared state but also lets us create files.

What about something like:
```python
def setUp(self):
    super().setUp()
    self.file_path = 'test_%s.conf' % uuid.uuid4()
```

and then leave tests as they were. That would ensure a unique file_name for each test. The only downside with that approach is there would need to be a cleanup in each test to delete the file for each test (but that was already done before).

**Review by mtreinish — COMMENTED** [review]
Thanks a bunch for doing this, I think it's definitely a good step to move to method level parallelism for CI. I do wonder if we'll encounter issues with 2 test methods in a class conflicting with each other (even for just limited resources) but we can tackle that when the time comes. 

I left a comment inline about a potential alternative approach for user config tests, I'm personally not a fan of the splitting the config for the tests out into separate files. Other than that it doesn't look `.stestr.conf` is actually modified in this PR, was your intent to remove the `group_regex` flag in this PR?

File: test/python/test_user_config.py (hunk 1 above)
Comment (jorconnor) [reply test/python/test_user_config.py]: Oh doh. Thanks, I am unsure why the grouping change wasn't included in that commit. I must have done something funky when I rebased on main.

Yea I considered using a uuid to create a filename. My gut was sort of against it, but I do not  really have any specific reasons for it. I think I just have a hesitation of writing files in a test unless it is necessary.   If you are fine with the uuid then that is enough for me to get over it though.

**Review by mtreinish — APPROVED** [review]
LGTM, thanks for the quick update

**Pull request merged**